src/hal/src/pso/descriptor.rs

```diff
@@ -33,38 +33,67 @@ pub type DescriptorBinding = u32;
 ///
 pub type DescriptorArrayIndex = usize;
 
-/// DOC TODO: Grasping and remembering the differences between these
-///       types is a tough task. We might be able to come up with better names?
-///       Or even use tuples to describe functionality instead of coming up with fancy names.
-#[repr(C)]
-#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+/// Specific type of a buffer.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub enum BufferDescriptorType {
+    /// Storage buffers allow load, store, and atomic operations.
+    Storage {
+        /// If true, store operations are not permitted on this buffer.
+        read_only: bool,
+    },
+    /// Uniform buffers provide constant data to be accessed in a shader.
+    Uniform,
+}
+
+/// Format of a buffer.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub enum BufferDescriptorFormat {
+    /// The buffer is interpreted as a structure defined in a shader.
+    Structured {
+        /// If true, the buffer is accessed by an additional offset specified in
+        /// the `offsets` parameter of `CommandBuffer::bind_*_descriptor_sets`.
+        dynamic_offset: bool,
+    },
+    /// The buffer is interpreted as a 1-D array of texels, which undergo format
+    /// conversion when loaded in a shader.
+    Texel,
+}
+
+/// Specific type of an image descriptor.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub enum ImageDescriptorType {
+    /// A sampled image allows sampling operations.
+    Sampled,
```

Comment:
Would you be open to have this as :
```rust
Sampled { with_sampler: bool },
```
Such that it replaces `CombinedImageSampler`.

The reasoning is for #3055, as the description says, we'll need to add a few fields to each image descriptor:
  - dimensionality
  - multisampling bit

It would be much more convenient for us and the users if there was a single entry point for image descriptors, even ones that come combined with samplers.